compiler/rustc_middle/src/query/mod.rs

```diff
@@ -1373,9 +1373,10 @@ rustc_queries! {
     /// Gets the ParameterEnvironment for a given item; this environment
     /// will be in "user-facing" mode, meaning that it is suitable for
     /// type-checking etc, and it does not normalize specializable
-    /// associated types. This is almost always what you want,
-    /// unless you are doing MIR optimizations, in which case you
-    /// might want to use `reveal_all()` method to change modes.
+    /// associated types.
+    ///
+    /// You should pretty much only use this if an `infcx` is available,
+    /// otherwise use a `TypingEnv`.
```

Comment:
I think this could be stronger. Someting like:

> you should almost certainly not use this -- if you already have an `InferCtxt`, then you should also probably have a `ParamEnv` from when it was built. If you don't, then you should take a `TypingEnv` to ensure that you handle opaque types correctly.